### DISCUSSION TITLE
about "workspace/didChangeConfiguration"

### DISCUSSION THREAD

**Discussion by Yurii-huang** [clangd/clangd#1168] [Q&A] [answered]
the [](https://github.com/llvm/llvm-project/blob/d480f968ad8b56d3ee4a6b6df5532d485b0ad01e/clang-tools-extra/clangd/test/dependency-output.test)
code lens:
```
{"jsonrpc":"2.0","method":"workspace/didChangeConfiguration","params":{"settings":{"compilationDatabaseChanges":{"/clangd-test/foo.c":
{"workingDirectory":"/clangd-test", "compilationCommand": ["clang", "-c", "-Xclang", "--show-includes", "-Xclang", "-sys-header-deps", "foo.c"]}}}}}
```

and from LSP protocol:
```
interface DidChangeConfigurationParams {
	/**
	 * The actual changed settings
	 */
	settings: LSPAny;
}
```

about "workspace/didChangeConfiguration".
I can't find the relevant usage explanation in the LSP protocol, but the current instruction is defined as any type by the standard protocol. 
In the Clangd document, it seems that the completionDatabase (remain is JSON file) can be changed through this instruction, and then the compiled database can be updated through this instruction, right? 
If a local file is currently deleted, do you also need to notify Clangd of such an operation?

**Comment by HighCommander4** [comment, marked as answer]
> about "workspace/didChangeConfiguration". I can't find the relevant usage explanation in the LSP protocol, but the current instruction is defined as any type by the standard protocol.

I believe that's because the protocol doesn't define any standard format or semantics for this message. It's up to each language server to decide what kind of configuration can be updated with this message, and what kind of data it expects inside `settings`.

> In the Clangd document, it seems that the completionDatabase (remain is JSON file) can be changed through this instruction, and then the compiled database can be updated through this instruction, right? If a local file is currently deleted, do you also need to notify Clangd of such an operation?

The purpose of this functionality is to **override** the compile command for a source file, to something different than what is specified in the `compile_commands.json` (or to set a compile command for a file which is missing from the `compile_commands.json`). There is no need for the client to notify clangd of changes to the `compile_commands.json` file itself (including deletion); clangd watches that file for changes automatically.

**Reply by Yurii-huang** [reply]
Thank you for your reply. I got a positive reply from it.

> There is no need for the client to notify clangd of changes to the compile_commands.json file itself (including deletion); 

I have another question.
Is the description of "compile_commands.json" equivalent to the concept of "workspace" of language protocol, or is it a description file of workspace?

**Reply by Yurii-huang** [reply]
> There is no need for the client to notify clangd of changes to the compile_commands.json file itself (including deletion); 

If I do not delete but I change a source file, I need to use Workspace/didchangeconfiguration refresh the compiled database?

**Reply by HighCommander4** [reply]
> I have another question. Is the description of "compile_commands.json" equivalent to the concept of "workspace" of language protocol, or is it a description file of workspace?

The `compile_commands.json` (also called "compilation database") contains a list of source files in the project, and commands that tell clangd how to compile each file.

The format is not specific to clangd, it's used by other clang-based tools (e.g. clang-tidy) as well. More info can be found at https://clang.llvm.org/docs/JSONCompilationDatabase.html.

**Reply by HighCommander4** [reply]
> > There is no need for the client to notify clangd of changes to the compile_commands.json file itself (including deletion);
> 
> If I do not delete but I change a source file, I need to use Workspace/didchangeconfiguration refresh the compiled database?

If you change a source file's entry in the compilation database, there's no need to send any LSP message, clangd will detect the change by itself (it watches the file for changes).

**Reply by Yurii-huang** [reply]
> There is no need for the client to notify clangd of changes to the compile_commands.json file itself (including deletion);
> 
> If I do not delete but I change a source file, I need to use Workspace/didchangeconfiguration refresh the compiled database?
> 
> If you change a source file's entry in the compilation database, there's no need to send any LSP message, clangd will detect the change by itself (it watches the file for changes).

Thank well.
I understand that "complie_command.json" only describes the rules between language source files.
The "workspace" in LSP protocol tends to be an abstract concept. It can describe Compilation Rules and more rules, although it is usually represented by folder.
